containertest/README.md

```diff
@@ -0,0 +1,112 @@
+# Mysqltest library
+
+## Introduction
+
+This is a Go library for starting an ephemeral Docker container. It is used to test
+integration between Go code and services such as PostgreSQL.
+
+## How to use it
+
+For some containers such as DBs, startup can take a while. You may want to reuse
+the same container and `USE` separate databases/namespaces or separate tables to
+isolate your tests from each other.
+
+Call `MustStart()` from your `TestMain()` function, If you're not familiar with Go's `TestMain()`
+mechanism for global test initialization, see the docs: https://pkg.go.dev/testing#hdr-Main.
+
+MySQL will be used as an example service, though any Service implementation could
+be used. Currently two implementations of `containertest.Service` exist in this
+repo, stored in `mysql.go` and `postgres.go`.
+
+```go
+package mypackage
+
+import (
+    "database/sql"
+    "fmt"
+    "io"
+    "net"
+    "os"
+    "testing"
+
+    "github.com/abcxyz/pkg/containertest"
+    _ "github.com/go-sql-driver/mysql" // Link with the Go MySQL driver
+)
+
+var ci containertest.ConnInfo
+var mySQLService *containertest.MySQL
+
+// TestMain runs once at startup to do test initialization common to all tests.
+func TestMain(m *testing.M) {
+    var closer io.Closer
+
+	mySQLService := (&MySQL{}).WithVersion("8.0")
+    ci, closer = containertest.MustStart(mySQLService) // Start the docker container. Can also pass options.
+    exitCode := m.Run() // Runs unit tests
+
+    // Remove container. If tests panic, this won't run, but there's nothing we can do about that:
+    // https://github.com/golang/go/issues/37206#issuecomment-590441512. In that case, then the
+    // container will eventually time out and be cleaned up. Defer isn't used because
+	// os.Exit() will prevent execution, and is needed for TestMain pattern.
+    closer.Close()
```

Comment:
I don't think it's very idiomatic to return a closer _structure_. Usually a closer is a function (two return values) or dangles on the returned object (one return value).

Reply:
I was following the existing pattern.

Would it make more sense for closer to be a function field in the `ConnInfo` struct which would have a `Close()` method that calls it?


```go
// ConnInfo specifies how connect to the created container.
type ConnInfo struct {
	Host string

	// PortMapper maps from container port to host port. Do not use after container is closed.
	PortMapper func(string) string
	
	closer *containerCloser
}

func (c *ConnInfo) Close() error {
	return c.closer.Close();
}

```

Reply:
Or possibly
```go
// ConnInfo specifies how connect to the created container.
type ConnInfo struct {
	Host string

	// PortMapper maps from container port to host port. Do not use after container is closed.
	PortMapper func(string) string
	
	closer func() error
}

func (c *ConnInfo) Close() error {
	return c.closer();
}
```

Reply:
The internal implementation isn't super important, but it would be most common to have this dangle off the ConnInfo structure as a function like you have in both examples.